runtime/gc_ctrl.c

```diff
@@ -197,13 +199,31 @@ CAMLprim value caml_gc_set(value v)
   /* Minor heap size comes last because it will trigger a minor collection
      (thus invalidating [v]) and it can raise [Out_of_memory]. */
   newminwsz = caml_norm_minor_heap_size (Long_val (Field (v, 0)));
-  if (newminwsz != Caml_state->minor_heap_wsz){
+
+  if (newminwsz != Caml_state->minor_heap_wsz) {
     caml_gc_message (0x20, "New minor heap size: %"
                      ARCH_INTNAT_PRINTF_FORMAT "uk words\n", newminwsz / 1024);
+  }
+
+  if (newminwsz > caml_minor_heap_max_wsz) {
+    caml_gc_log ("update minor heap max: %"
+                 ARCH_SIZET_PRINTF_FORMAT "uk words", newminwsz / 1024);
+    caml_update_minor_heap_max(newminwsz);
+  }
+  CAMLassert(newminwsz <= caml_minor_heap_max_wsz);
+  if (newminwsz != Caml_state->minor_heap_wsz) {
+    caml_gc_log ("current minor heap size: %"
+                 ARCH_SIZET_PRINTF_FORMAT "uk words",
+                 Caml_state->minor_heap_wsz / 1024);
+    caml_gc_log ("set minor heap size: %"
+                 ARCH_SIZET_PRINTF_FORMAT "uk words", newminwsz / 1024);
+    /* FIXME: when (newminwsz > caml_minor_heap_max_wsz) and
```

Comment:
Is there a reason not to move the functionality of `caml_update_minor_heap_max` inside `caml_set_minor_heap_size`?
Wouldn't that deal with the `FIXME` and also reduce the surface of functions to just `caml_set_minor_heap_size`.

Reply:
Agreed, there's no other entrypoint to `caml_set_minor_heap_size` (nor to `caml_update_minor_heap_max`) so this wouldn't add any unwanted feature to either functions.

Reply:
Moving the functionality is a good suggestion, but I believe that it would not make the FIXME disappear.

When we update the minor heap max, we need to change the minor heaps area and then each domain then re-commits its own minor-heap memory -- towards the end of the STW section. In particular, the current domain (which is planning to change its minor heap size) recommits its "previous" minor heap size. And then we run `caml_set_minor_heap_size` which will decommit and recommit again; there is a duplication here (the domain decommits and recommits twice) which is the FIXME.

If we simply move `caml_set_minor_heap_size` into `caml_update_minor_heap_max`, the duplciation will still be present. If we wanted to fix the FIXME, we would need to pass the information about the desired *new* heap size to the STW section. But this would use a sensibly more complex API: we need to tell each domain which size to use (with most domains not changing their own size... except if we later implement a different "change the minor heap size of all domains" knob). I think it is wiser to wait for the "change the maximum number of domains" feature set to land (as further PRs), and see what API we have at this point, to revisit this question.

Note: one thing we tried is to update `Caml_state->minor_heap_wsz` before calling `caml_update_minor_heap_max`, so that when the domain recommits its minor heap, it naturally uses the new setting. But I don't trust this way of doing things, because it means that between the update and the recommit the domain is in a slightly inconsistent state, with a `minor_heap_wsz` setting that is inconsistent with its actual `minor_young_end - minor_young_start` span. This could work if we audit the code carefully to ensure that we never rely on `minor_heap_wsz` (except when committing new memory), but the code running between the update and the recommit contains a lot of subtle GC functions and I'm not sure about adding new delicate constraints to them. We could also consider having two values in the domain state, "current minor heap size" and "desired minor heap size", and have the recommit use the "desired" information.

I'm not sure that the complexity of dealing with this FIXME is worth it. Maybe the best choice is to keep the code somewhat naive, with the redundant decommit/recommit, knowing that this operation is very rare anyway.

Reply:
I think that moving `caml_update_minor_heap_max` into `caml_set_minor_heap_size` could work:

If we can easily identify the domain that initiated the STW section (which, if I understand correctly, we can, since it is `participating[0]`), our protocol would be:

In `caml_set_minor_heap_size`, the domain frees its own heap before initiating the STW section.

During the STW section,

1. all domains except `participating[0]` free their minor heap.
2. As previously, one domain modifies all domains' minor heap areas during the STW section.
3. all domains except `participating[0]` allocate their minor heap.

After the STW section, back in `caml_set_minor_heap_size`, the domain allocates its own own heap in the desired new size.

Does that make sense? I'll implement and see.

Reply:
(I think it is not `participating[0]`, which is likely to be domain-0, but rather `stw_leader`, but I'm not sure.)

I think we want to use the same STW code to deal with Max_domains update in the near future. In the case where we want to update Max_domains, we want a different protocol because the current domain does not play a special role. Similarly, if we implement a "resize the minor heap for all active domains" entry point in Gc, we need a different interface again. Presonally I would be in favor of replacing the "Max_domains is a constant" part, which is a more severe limitation than this FIXME, before we look at what the STW interface is and whether we want to add more stuff to optionally change the leader's minor heap size.

Reply:
You're right, it's not `participating[0]`.

For max_domains, I am not 100% sure that changing this limit at runtime has a use case that warrants taking on the additional implementation complexity. It looks to me like

a. keeping the default number of max_domains at 128 (which will work for the vast majority of architectures), and
b. introducing an OCAMLRUNPARAM to set max_domains

is a reasonable alternative that would only require us to change the initialization code of the domain-indexed data structures.

Still, implementing the ability to change max_domains via `Gc.set` seems interesting in terms of identifying invariants and current limitations of some parts of the runtime.

Reply:
Identifying the domain who initiated the STW section via `((dom_internal *) atomic_load_acq(&stw_leader))->state`, and making this domain responsible for allocating its own heap in `caml_set_minor_heap_size`, after the STW section seems to work, the tests pass.

One open question: if we do make max_domains changeable at runtime, should that live on `Gc.set` or on the `Domain` interface?

Reply:
In the change_minor_heap_size_FIXME branch https://github.com/sabine/ocaml/commit/f37be6ba8b56e78a60f052199f13d4748ea9352d:

The domain on which `Gc.set` was called initiates the STW section if the requested size is bigger than the current reservation.

During the STW section,

    all domains free their minor heap.
    As previously, one domain modifies all domains' minor heap areas during the STW section.
    all domains except the initiating domain (stw_leader) allocate their minor heap.

After the STW section, back in caml_set_minor_heap_size, the STW-initiating domain allocates its own own heap in the desired new size.

Reply:
Can we leave this FIXME in the current PR? Doing it like this, with the spurious reallocation of the leader domain's minor heap was the least invasive change. This patch does the heavy lifting of introducing a STW section to deallocate/unreserve/reserve/allocate the minor heaps.

A follow-up PR that resolves the FIXME by adjusting the semantics of `Gc.set` to change the minor heap size of all domains (with resulting simplifications in the codebase) is mostly ready.